File: packages/less/src/less/functions/math-helper.test.js
import { describe, it, expect } from 'vitest';
import MathHelper from './math-helper.js';
import Dimension from '../tree/dimension.js';
import Unit from '../tree/unit.js';

describe('MathHelper', () => {
    describe('argument validation', () => {
        it('should throw error for non-Dimension argument', () => {
            const fn = x => x * 2;
            
            expect(() => MathHelper(fn, null, 'not-a-dimension')).toThrow();
            expect(() => MathHelper(fn, null, 42)).toThrow();
            expect(() => MathHelper(fn, null, null)).toThrow();
            expect(() => MathHelper(fn, null, undefined)).toThrow();
            expect(() => MathHelper(fn, null, {})).toThrow();
        });

        it('should throw error with specific message for invalid argument', () => {
            const fn = x => x * 2;
            
            try {
                MathHelper(fn, null, 'invalid');
                expect.fail('Expected error to be thrown');
            } catch (error) {
                expect(error.type).toBe('Argument');
                expect(error.message).toBe('argument must be a number');
            }
        });

        it('should accept valid Dimension argument', () => {
            const fn = x => x * 2;
            const dim = new Dimension(5);
            
            expect(() => MathHelper(fn, null, dim)).not.toThrow();
        });
    });

    describe('unit handling', () => {
        it('should use dimension unit when unit parameter is null', () => {
            const fn = x => x * 2;
            const unit = new Unit(['px'], [], 'px');
            const dim = new Dimension(5, unit);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.unit).toBe(unit);
            expect(result.value).toBe(10);
        });

        it('should use dimension unit when unit parameter is undefined', () => {
            const fn = x => x * 2;
            const unit = new Unit(['em'], [], 'em');
            const dim = new Dimension(3, unit);
            
            const result = MathHelper(fn, undefined, dim);
            
            // JavaScript treats undefined as falsy, so it calls unify
            expect(result.value).toBe(6);
        });

        it('should use provided unit when unit parameter is specified', () => {
            const fn = x => x * 2;
            const originalUnit = new Unit(['px'], [], 'px');
            const newUnit = new Unit(['em'], [], 'em');
            const dim = new Dimension(5, originalUnit);
            
            // Mock the unify method to return a unified dimension
            dim.unify = () => new Dimension(5, originalUnit);
            
            const result = MathHelper(fn, newUnit, dim);
            
            expect(result.unit).toBe(newUnit);
            expect(result.value).toBe(10);
        });

        it('should call unify when unit parameter is provided', () => {
            const fn = x => x * 2;
            const originalUnit = new Unit(['px'], [], 'px');
            const newUnit = new Unit(['em'], [], 'em');
            const dim = new Dimension(5, originalUnit);
            
            let unifyCalled = false;
            dim.unify = () => {
                unifyCalled = true;
                return new Dimension(5, originalUnit);
            };
            
            MathHelper(fn, newUnit, dim);
            
            expect(unifyCalled).toBe(true);
        });

        it('should handle unitless dimensions', () => {
            const fn = x => x * 3;
            const dim = new Dimension(4);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(12);
            expect(result.unit).toBe(dim.unit);
        });
    });

    describe('mathematical operations', () => {
        it('should apply simple multiplication function', () => {
            const fn = x => x * 2;
            const dim = new Dimension(5);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(10);
        });

        it('should apply simple addition function', () => {
            const fn = x => x + 10;
            const dim = new Dimension(5);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(15);
        });

        it('should apply division function', () => {
            const fn = x => x / 2;
            const dim = new Dimension(10);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(5);
        });

        it('should apply subtraction function', () => {
            const fn = x => x - 3;
            const dim = new Dimension(8);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(5);
        });

        it('should apply power function', () => {
            const fn = x => Math.pow(x, 2);
            const dim = new Dimension(4);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(16);
        });

        it('should apply square root function', () => {
            const fn = x => Math.sqrt(x);
            const dim = new Dimension(16);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(4);
        });

        it('should apply trigonometric functions', () => {
            const sinFn = x => Math.sin(x);
            const dim = new Dimension(Math.PI / 2);
            
            const result = MathHelper(sinFn, null, dim);
            
            expect(result.value).toBeCloseTo(1, 10);
        });

        it('should handle negative numbers', () => {
            const fn = x => Math.abs(x);
            const dim = new Dimension(-5);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(5);
        });

        it('should handle decimal numbers', () => {
            const fn = x => x * 2;
            const dim = new Dimension(3.14);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBeCloseTo(6.28, 10);
        });

        it('should handle zero values', () => {
            const fn = x => x + 1;
            const dim = new Dimension(0);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(1);
        });

        it('should handle very large numbers', () => {
            const fn = x => x / 1000000;
            const dim = new Dimension(1000000000);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(1000);
        });

        it('should handle very small numbers', () => {
            const fn = x => x * 1000;
            const dim = new Dimension(0.001);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(1);
        });
    });

    describe('return value properties', () => {
        it('should return a Dimension instance', () => {
            const fn = x => x * 2;
            const dim = new Dimension(5);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result).toBeInstanceOf(Dimension);
        });

        it('should preserve unit when no unit conversion', () => {
            const fn = x => x * 2;
            const unit = new Unit(['px'], [], 'px');
            const dim = new Dimension(5, unit);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.unit).toBe(unit);
        });

        it('should use specified unit when provided', () => {
            const fn = x => x * 2;
            const originalUnit = new Unit(['px'], [], 'px');
            const newUnit = new Unit(['em'], [], 'em');
            const dim = new Dimension(5, originalUnit);
            
            // Mock unify to return the dimension unchanged
            dim.unify = () => dim;
            
            const result = MathHelper(fn, newUnit, dim);
            
            expect(result.unit).toBe(newUnit);
        });
    });

    describe('complex scenarios', () => {
        it('should handle chained mathematical operations', () => {
            const fn1 = x => x * 2;
            const fn2 = x => x + 5;
            const dim = new Dimension(3);
            
            const result1 = MathHelper(fn1, null, dim);
            const result2 = MathHelper(fn2, null, result1);
            
            expect(result2.value).toBe(11); // (3 * 2) + 5 = 11
        });

        it('should work with complex units', () => {
            const fn = x => x * 2;
            const unit = new Unit(['px', 'em'], ['s'], 'px*em/s');
            const dim = new Dimension(5, unit);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(10);
            expect(result.unit).toBe(unit);
        });

        it('should handle function that returns NaN', () => {
            const fn = x => Math.sqrt(-x); // Will return NaN for positive x
            const dim = new Dimension(5);
            
            // This will throw because Dimension constructor rejects NaN values
            expect(() => MathHelper(fn, null, dim)).toThrow('Dimension is not a number.');
        });

        it('should handle function that returns Infinity', () => {
            const fn = x => 1 / x;
            const dim = new Dimension(0);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(Infinity);
        });

        it('should handle function that returns negative Infinity', () => {
            const fn = x => -1 / x;
            const dim = new Dimension(0);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(-Infinity);
        });
    });

    describe('parseFloat behavior', () => {
        it('should handle string values in dimension', () => {
            const fn = x => x * 2;
            const dim = new Dimension('5');
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(10);
        });

        it('should handle string values with decimals', () => {
            const fn = x => x * 2;
            const dim = new Dimension('3.14');
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBeCloseTo(6.28, 10);
        });

        it('should handle edge cases of parseFloat', () => {
            const fn = x => x + 1;
            
            // Test with various parseFloat edge cases
            expect(MathHelper(fn, null, new Dimension('5.5')).value).toBe(6.5);
            expect(MathHelper(fn, null, new Dimension('  10  ')).value).toBe(11);
            expect(MathHelper(fn, null, new Dimension('5.5abc')).value).toBe(6.5);
        });
    });

    describe('dimension with unify method', () => {
        it('should use the dimension returned by unify when unit is specified', () => {
            const fn = x => x * 2;
            const originalUnit = new Unit(['px'], [], 'px');
            const newUnit = new Unit(['em'], [], 'em');
            const dim = new Dimension(5, originalUnit);
            const unifiedDim = new Dimension(10, originalUnit); // Different value after unify
            
            dim.unify = () => unifiedDim;
            
            const result = MathHelper(fn, newUnit, dim);
            
            expect(result.value).toBe(20); // 10 * 2, using unified dimension value
            expect(result.unit).toBe(newUnit);
        });

        it('should not call unify when unit is null', () => {
            const fn = x => x * 2;
            const dim = new Dimension(5);
            
            let unifyCalled = false;
            dim.unify = () => {
                unifyCalled = true;
                return dim;
            };
            
            MathHelper(fn, null, dim);
            
            expect(unifyCalled).toBe(false);
        });

        it('should call unify when unit is undefined', () => {
            const fn = x => x * 2;
            const dim = new Dimension(5);
            
            let unifyCalled = false;
            dim.unify = () => {
                unifyCalled = true;
                return dim;
            };
            
            MathHelper(fn, undefined, dim);
            
            // JavaScript treats undefined as truthy in `if (unit === null)` check, so unify is called
            expect(unifyCalled).toBe(true);
        });
    });

    describe('edge cases and error conditions', () => {
        it('should handle empty function (identity)', () => {
            const fn = x => x;
            const dim = new Dimension(42);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(42);
        });

        it('should handle function that ignores input', () => {
            const fn = () => 100;
            const dim = new Dimension(42);
            
            const result = MathHelper(fn, null, dim);
            
            expect(result.value).toBe(100);
        });

        it('should work with falsy unit values', () => {
            const fn = x => x * 2;
            const dim = new Dimension(5);
            
            const result1 = MathHelper(fn, null, dim);
            const result2 = MathHelper(fn, undefined, dim);
            const result3 = MathHelper(fn, 0, dim);
            
            expect(result1.value).toBe(10);
            expect(result2.value).toBe(10);
            expect(result3.value).toBe(10);
        });
    });
});

File: packages/less/src/less/less_go/math_helper_test.go
package less_go

import (
	"math"
	"testing"
)

func TestMathHelper(t *testing.T) {
	t.Run("argument validation", func(t *testing.T) {
		fn := func(x float64) float64 { return x * 2 }

		t.Run("should return error for non-Dimension argument", func(t *testing.T) {
			testCases := []any{"not-a-dimension", 42, nil, map[string]any{}}
			
			for _, testCase := range testCases {
				_, err := MathHelper(fn, nil, testCase)
				if err == nil {
					t.Errorf("Expected error for input %v, but got none", testCase)
				}
				
				if mathErr, ok := err.(*MathHelperError); ok {
					if mathErr.Type != "Argument" || mathErr.Message != "argument must be a number" {
						t.Errorf("Expected specific error message, got: %v", err)
					}
				} else {
					t.Errorf("Expected MathHelperError, got: %T", err)
				}
			}
		})

		t.Run("should accept valid Dimension argument", func(t *testing.T) {
			dim, _ := NewDimension(5.0, nil)
			_, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Errorf("Expected no error for valid Dimension, got: %v", err)
			}
		})
	})

	t.Run("unit handling", func(t *testing.T) {
		fn := func(x float64) float64 { return x * 2 }

		t.Run("should use dimension unit when unit parameter is nil", func(t *testing.T) {
			unit := NewUnit([]string{"px"}, nil, "px")
			dim, _ := NewDimension(5.0, unit)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Unit != unit {
				t.Errorf("Expected same unit, got different")
			}
			if result.Value != 10 {
				t.Errorf("Expected value 10, got %f", result.Value)
			}
		})

		t.Run("should use provided unit when unit parameter is specified", func(t *testing.T) {
			originalUnit := NewUnit([]string{"px"}, nil, "px")
			newUnit := NewUnit([]string{"em"}, nil, "em")
			dim, _ := NewDimension(5.0, originalUnit)
			
			result, err := MathHelper(fn, newUnit, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Unit != newUnit {
				t.Errorf("Expected new unit, got original unit")
			}
			if result.Value != 10 {
				t.Errorf("Expected value 10, got %f", result.Value)
			}
		})

		t.Run("should handle unitless dimensions", func(t *testing.T) {
			dim, _ := NewDimension(4.0, nil)
			
			result, err := MathHelper(func(x float64) float64 { return x * 3 }, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 12 {
				t.Errorf("Expected value 12, got %f", result.Value)
			}
		})
	})

	t.Run("mathematical operations", func(t *testing.T) {
		t.Run("should apply simple multiplication function", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 2 }
			dim, _ := NewDimension(5.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 10 {
				t.Errorf("Expected value 10, got %f", result.Value)
			}
		})

		t.Run("should apply simple addition function", func(t *testing.T) {
			fn := func(x float64) float64 { return x + 10 }
			dim, _ := NewDimension(5.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 15 {
				t.Errorf("Expected value 15, got %f", result.Value)
			}
		})

		t.Run("should apply division function", func(t *testing.T) {
			fn := func(x float64) float64 { return x / 2 }
			dim, _ := NewDimension(10.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 5 {
				t.Errorf("Expected value 5, got %f", result.Value)
			}
		})

		t.Run("should apply subtraction function", func(t *testing.T) {
			fn := func(x float64) float64 { return x - 3 }
			dim, _ := NewDimension(8.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 5 {
				t.Errorf("Expected value 5, got %f", result.Value)
			}
		})

		t.Run("should apply power function", func(t *testing.T) {
			fn := func(x float64) float64 { return math.Pow(x, 2) }
			dim, _ := NewDimension(4.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 16 {
				t.Errorf("Expected value 16, got %f", result.Value)
			}
		})

		t.Run("should apply square root function", func(t *testing.T) {
			fn := func(x float64) float64 { return math.Sqrt(x) }
			dim, _ := NewDimension(16.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 4 {
				t.Errorf("Expected value 4, got %f", result.Value)
			}
		})

		t.Run("should apply trigonometric functions", func(t *testing.T) {
			fn := func(x float64) float64 { return math.Sin(x) }
			dim, _ := NewDimension(math.Pi/2, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if math.Abs(result.Value-1.0) > 1e-10 {
				t.Errorf("Expected value close to 1, got %f", result.Value)
			}
		})

		t.Run("should handle negative numbers", func(t *testing.T) {
			fn := func(x float64) float64 { return math.Abs(x) }
			dim, _ := NewDimension(-5.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 5 {
				t.Errorf("Expected value 5, got %f", result.Value)
			}
		})

		t.Run("should handle decimal numbers", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 2 }
			dim, _ := NewDimension(3.14, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if math.Abs(result.Value-6.28) > 1e-10 {
				t.Errorf("Expected value close to 6.28, got %f", result.Value)
			}
		})

		t.Run("should handle zero values", func(t *testing.T) {
			fn := func(x float64) float64 { return x + 1 }
			dim, _ := NewDimension(0.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 1 {
				t.Errorf("Expected value 1, got %f", result.Value)
			}
		})

		t.Run("should handle very large numbers", func(t *testing.T) {
			fn := func(x float64) float64 { return x / 1000000 }
			dim, _ := NewDimension(1000000000.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 1000 {
				t.Errorf("Expected value 1000, got %f", result.Value)
			}
		})

		t.Run("should handle very small numbers", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 1000 }
			dim, _ := NewDimension(0.001, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 1 {
				t.Errorf("Expected value 1, got %f", result.Value)
			}
		})
	})

	t.Run("return value properties", func(t *testing.T) {
		t.Run("should return a Dimension instance", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 2 }
			dim, _ := NewDimension(5.0, nil)
			
			_, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			// MathHelper returns *Dimension by design, no need for additional checks
		})

		t.Run("should preserve unit when no unit conversion", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 2 }
			unit := NewUnit([]string{"px"}, nil, "px")
			dim, _ := NewDimension(5.0, unit)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Unit != unit {
				t.Errorf("Expected same unit object, got different")
			}
		})

		t.Run("should use specified unit when provided", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 2 }
			originalUnit := NewUnit([]string{"px"}, nil, "px")
			newUnit := NewUnit([]string{"em"}, nil, "em")
			dim, _ := NewDimension(5.0, originalUnit)
			
			result, err := MathHelper(fn, newUnit, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Unit != newUnit {
				t.Errorf("Expected new unit, got %v", result.Unit)
			}
		})
	})

	t.Run("complex scenarios", func(t *testing.T) {
		t.Run("should handle chained mathematical operations", func(t *testing.T) {
			fn1 := func(x float64) float64 { return x * 2 }
			fn2 := func(x float64) float64 { return x + 5 }
			dim, _ := NewDimension(3.0, nil)
			
			result1, err := MathHelper(fn1, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error in first operation: %v", err)
			}
			
			result2, err := MathHelper(fn2, nil, result1)
			if err != nil {
				t.Fatalf("Unexpected error in second operation: %v", err)
			}
			
			if result2.Value != 11 { // (3 * 2) + 5 = 11
				t.Errorf("Expected value 11, got %f", result2.Value)
			}
		})

		t.Run("should work with complex units", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 2 }
			unit := NewUnit([]string{"px", "em"}, []string{"s"}, "px*em/s")
			dim, _ := NewDimension(5.0, unit)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 10 {
				t.Errorf("Expected value 10, got %f", result.Value)
			}
			if result.Unit != unit {
				t.Errorf("Expected same complex unit, got different")
			}
		})

		t.Run("should handle function that returns NaN", func(t *testing.T) {
			fn := func(x float64) float64 { return math.Sqrt(-x) } // Will return NaN for positive x
			dim, _ := NewDimension(5.0, nil)
			
			_, err := MathHelper(fn, nil, dim)
			
			// This should error because NewDimension rejects NaN values
			if err == nil {
				t.Error("Expected error for NaN result, but got none")
			}
		})

		t.Run("should handle function that returns Infinity", func(t *testing.T) {
			fn := func(x float64) float64 { return 1.0 / x }
			dim, _ := NewDimension(0.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if !math.IsInf(result.Value, 1) {
				t.Errorf("Expected +Inf, got %f", result.Value)
			}
		})

		t.Run("should handle function that returns negative Infinity", func(t *testing.T) {
			fn := func(x float64) float64 { return -1.0 / x }
			dim, _ := NewDimension(0.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if !math.IsInf(result.Value, -1) {
				t.Errorf("Expected -Inf, got %f", result.Value)
			}
		})
	})

	t.Run("edge cases and error conditions", func(t *testing.T) {
		t.Run("should handle empty function (identity)", func(t *testing.T) {
			fn := func(x float64) float64 { return x }
			dim, _ := NewDimension(42.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 42 {
				t.Errorf("Expected value 42, got %f", result.Value)
			}
		})

		t.Run("should handle function that ignores input", func(t *testing.T) {
			fn := func(x float64) float64 { return 100 }
			dim, _ := NewDimension(42.0, nil)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			if result.Value != 100 {
				t.Errorf("Expected value 100, got %f", result.Value)
			}
		})
	})

	t.Run("unify behavior", func(t *testing.T) {
		t.Run("should call Unify when unit is provided", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 2 }
			originalUnit := NewUnit([]string{"px"}, nil, "px")
			newUnit := NewUnit([]string{"em"}, nil, "em")
			dim, _ := NewDimension(5.0, originalUnit)
			
			result, err := MathHelper(fn, newUnit, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			// After unify, the value should be the same (since we don't have real conversion)
			// but the final unit should be the provided unit
			if result.Value != 10 { // 5 * 2 (unify doesn't change value in this simple case)
				t.Errorf("Expected value 10, got %f", result.Value)
			}
			if result.Unit != newUnit {
				t.Errorf("Expected new unit, got %v", result.Unit)
			}
		})

		t.Run("should not call Unify when unit is nil", func(t *testing.T) {
			fn := func(x float64) float64 { return x * 2 }
			unit := NewUnit([]string{"px"}, nil, "px")
			dim, _ := NewDimension(5.0, unit)
			
			result, err := MathHelper(fn, nil, dim)
			if err != nil {
				t.Fatalf("Unexpected error: %v", err)
			}
			
			// Should use original unit without calling Unify
			if result.Unit != unit {
				t.Errorf("Expected original unit, got different")
			}
		})
	})
}